library/common/extensions/filters/http/platform_bridge/filter.cc

```diff
@@ -562,9 +503,70 @@ void PlatformBridgeFilter::FilterBase::onResume(Buffer::Instance* internal_buffe
     replaceHeaders(*pending_trailers_, *result.pending_trailers);
     pending_trailers_ = nullptr;
     free(result.pending_trailers);
+  } else if (result.pending_trailers) {
+    addTrailers(*result.pending_trailers);
   }
+
   iteration_state_ = IterationState::Ongoing;
-  resume_call();
+  resumeIteration();
+}
+
+void PlatformBridgeFilter::RequestFilterBase::addData(envoy_data data) {
+  Buffer::OwnedImpl inject_data;
+  inject_data.addBufferFragment(*Buffer::BridgeFragment::createBridgeFragment(data));
+  parent_.decoder_callbacks_->addDecodedData(inject_data, /* watermark */ false);
+}
+
+void PlatformBridgeFilter::ResponseFilterBase::addData(envoy_data data) {
+  Buffer::OwnedImpl inject_data;
+  inject_data.addBufferFragment(*Buffer::BridgeFragment::createBridgeFragment(data));
+  parent_.encoder_callbacks_->addEncodedData(inject_data, /* watermark */ false);
+}
+
+void PlatformBridgeFilter::RequestFilterBase::addTrailers(envoy_headers trailers) {
+  Http::HeaderMap& inject_trailers = parent_.decoder_callbacks_->addDecodedTrailers();
+  replaceHeaders(inject_trailers, trailers);
+}
+
+void PlatformBridgeFilter::ResponseFilterBase::addTrailers(envoy_headers trailers) {
+  Http::HeaderMap& inject_trailers = parent_.encoder_callbacks_->addEncodedTrailers();
+  replaceHeaders(inject_trailers, trailers);
+}
+
+void PlatformBridgeFilter::RequestFilterBase::resumeIteration() {
+  parent_.encoder_callbacks_->continueEncoding();
+}
+
+void PlatformBridgeFilter::ResponseFilterBase::resumeIteration() {
+  parent_.decoder_callbacks_->continueDecoding();
+}
+
+// Technically-speaking to align with Envoy's internal API this method should take
+// a closure to execute with the available buffer, but since we control all usage,
+// this shortcut works for now.
```

Comment:
Doesn't seem like there'd be a reason to ever change this unless we _wanted_ to support the platform-layer modifying the data with a closure (which seems like extra complexity), no?

Reply:
I put this in more as a note to a future reader that we're actually doing something that might seem a little unexpected/wrong from the perspective of Envoy's API. The form of Envoy's API kind of implies (but doesn't state) that you should perform all modifications in the context of a closure rather than cheating and extracting a pointer to the buffer like we're doing. But since we only use it internally and we only use it in the scope of this call stack, it seems unlikely to ever be a problem, unless Envoy's filter pipeline and buffer management are dramatically overhauled. Tests would also reveal right away that this was broken if it ever changes.